webapp/src/modules/mapBuilder.js

```diff
@@ -75,104 +180,152 @@ export default class MapBuilder {
-      if (this._locationAccuracyCircle && this._locationMarker) {
-        this._animationTimeoutId = setTimeout(() => {
-          this._locationAccuracyCircle._path.classList.add(
-            "mobility-location-accuracy-circle-animation"
-          );
-          this._locationMarker._icon.style.transition = "all 1000ms linear 0s";
-        }, 250);
-      }
+      this._animationTimeoutId = setTimeout(() => {
+        this._locationAccuracyCircle._path.classList.add(
+          "mobility-location-accuracy-circle-animation"
+        );
+        this._locationMarker._icon.style.transition = "all 1000ms linear 0s";
+      }, 250);
     });
   }
 
-  setVehicleEventHandlers() {
-    this._map.on("moveend", this.moveEnd, this);
-    this._map.on("click", this.click, this);
+  _setVehicleEventHandlers() {
+    this._map.on("moveend", this._moveEndEvent, this);
+    this._map.on("click", this._clickEvent, this);
   }
 
-  moveEnd() {
-    if (!this._map) {
-      return;
-    }
+  _moveEndEvent() {
     const bounds = this._map.getBounds();
-    if (
-      !this._lastBounds.contains(bounds._northEast) ||
-      !this._lastBounds.contains(bounds._southWest)
-    ) {
-      this._lastBounds = bounds;
-      this.getScooters(bounds);
+    if (!this._map || this._lastBounds.contains(bounds)) {
+      return;
     }
+    this._lastBounds = bounds;
+    this._getAndUpdateScooters(this._lastBounds, this._mcg);
   }
 
-  click() {
-    if (this._vehicleClicked) {
-      this._onVehicleClick(null);
-      this._vehicleClicked = false;
+  _clickEvent() {
+    if (!this._vehicleClicked) {
+      return;
     }
+    this._onVehicleClick(null);
+    this._vehicleClicked = null;
   }
 
-  loadScooters({ onVehicleClick }) {
-    this._onVehicleClick = onVehicleClick;
-    this.getScooters(this._lastBounds);
-    this.setVehicleEventHandlers();
-    this._map.addLayer(this._mcg);
-    return this;
+  _getAndUpdateScooters(bounds, mcg) {
+    bounds = expandBounds(bounds, 0.05);
+    api.getMobilityMap(boundsToParams(bounds)).then((r) => {
+      this._updateMarkers({ ...r, bounds, mcg });
+      this._stopRefreshTimer()._startRefreshTimer(r.data.refresh);
+    });
   }
 
-  getScooters(bounds) {
-    api.getMobilityMap(boundsToParams(bounds)).then((r) => {
-      const precisionFactor = 1 / r.data.precision;
-      const newMarkers = [];
-      this.removeVisibleLayers(bounds);
-      ["ebike", "escooter"].forEach((vehicleType) => {
-        r.data[vehicleType]?.forEach((bike) => {
-          const marker = this.newMarker(
-            bike,
-            vehicleType,
-            r.data.providers,
-            precisionFactor
-          );
-          newMarkers.push(marker);
-        });
+  _updateMarkers({ data, bounds, mcg }) {
+    const precisionFactor = 1 / data.precision;
+    const applicableMarkers = [];
+    const allNewMarkersIds = [];
+    const leftoverMarkers = [];
+    // First: Removes markers that are not present in the bounds
```

Comment:
Reason for removing outside bound markers first is so we deal with less markers after

Reply:
Would love to optimize this code if possible